File: DesignPatternsInAutomatedTesting-Series/PatternsInAutomation.Tests/Advanced/Strategy/Basic/Base/PurchaseContext.cs
using PatternsInAutomatedTests.Advanced.Strategy.Data;
using PatternsInAutomatedTests.Advanced.Strategy.Pages.ItemPage;
using PatternsInAutomatedTests.Advanced.Strategy.Pages.PreviewShoppingCartPage;
using PatternsInAutomatedTests.Advanced.Strategy.Pages.ShippingAddressPage;
using PatternsInAutomatedTests.Advanced.Strategy.Pages.ShippingPaymentPage;
using PatternsInAutomatedTests.Advanced.Strategy.Pages.SignInPage;

namespace PatternsInAutomatedTests.Advanced.Strategy.Basic.Base
{
    public class PurchaseContext
    {
        private readonly IOrderValidationStrategy orderValidationStrategy;

        public PurchaseContext(IOrderValidationStrategy orderValidationStrategy)
        {
            this.orderValidationStrategy = orderValidationStrategy;
        }

        public void PurchaseItem(string itemUrl, string itemPrice, ClientLoginInfo clientLoginInfo, ClientPurchaseInfo clientPurchaseInfo)
        {
            ItemPage.Instance.Navigate(itemUrl);
            ItemPage.Instance.ClickBuyNowButton();
            PreviewShoppingCartPage.Instance.ClickProceedToCheckoutButton();
            SignInPage.Instance.Login(clientLoginInfo.Email, clientLoginInfo.Password);
            ShippingAddressPage.Instance.FillShippingInfo(clientPurchaseInfo);
            ShippingAddressPage.Instance.ClickContinueButton();
            ShippingPaymentPage.Instance.ClickBottomContinueButton();
            ShippingPaymentPage.Instance.ClickTopContinueButton();
            this.orderValidationStrategy.ValidateOrderSummary(itemPrice, clientPurchaseInfo);
        }
    }
}

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/advanced/base/PurchaseContext.java
package decorator.advanced.base;

import decorator.advanced.strategies.OrderPurchaseStrategy;
import decorator.core.Driver;
import decorator.data.PurchaseInfo;
import decorator.pages.checkoutpage.CheckoutPage;
import decorator.pages.itempage.ItemPage;
import decorator.pages.shoppingcartpage.ShoppingCartPage;

import java.math.BigDecimal;

public class PurchaseContext {
    private final OrderPurchaseStrategy orderPurchaseStrategy;

    public PurchaseContext(OrderPurchaseStrategy orderPurchaseStrategy) {
        this.orderPurchaseStrategy = orderPurchaseStrategy;
    }

    public void PurchaseItem(String itemUrl, BigDecimal itemPrice, PurchaseInfo clientPurchaseInfo) {
        Driver.getPage(ItemPage.class).open(itemUrl);
        Driver.getPage(ItemPage.class).clickBuyNowButton();
        Driver.getPage(ItemPage.class).clickViewShoppingCartButton();
        Driver.getPage(ShoppingCartPage.class).clickProceedToCheckoutButton();
        Driver.getPage(CheckoutPage.class).fillBillingInfo(clientPurchaseInfo);
        var expectedTotalPrice = orderPurchaseStrategy.calculateTotalPrice();
        orderPurchaseStrategy.validateOrderSummary(expectedTotalPrice);
    }
}
